plugins/alerting/pkg/alerting/api_logs.go

```diff
@@ -0,0 +1,176 @@
+	info []*alertingv1alpha.InformativeAlertLog
+	by   func(i1, i2 *alertingv1alpha.InformativeAlertLog) bool // Closure used in the Less method.
+}
+
+func (by By) Sort(alerts []*alertingv1alpha.InformativeAlertLog) {
+	alertInfoSorter := &alertInfoSorter{
+		info: alerts,
+		by:   by,
+	}
+	sort.Sort(alertInfoSorter)
+}
+
+// implement sort interface
+func (a *alertInfoSorter) Len() int {
+	return len(a.info)
+}
+
+func (a *alertInfoSorter) Swap(i, j int) {
+	a.info[i], a.info[j] = a.info[j], a.info[i]
+}
+
+func (a *alertInfoSorter) Less(i, j int) bool {
+	return a.by(a.info[i], a.info[j])
+}
+
+func (p *Plugin) CreateAlertLog(ctx context.Context, event *corev1.AlertLog) (*emptypb.Empty, error) {
+	if p.inMemCache != nil {
+		p.inMemCache.Add(event.ConditionId, event)
+	}
+	existing, err := GetIndices() // get the existing indices for each condition
+	if err != nil {
+		return nil, err
+	}
+	found := false
+	var b *BucketInfo
+	for _, index := range existing { // check if condition is already indexed
+		if index.ConditionId == event.ConditionId.Id {
+			found = true
+			b = index
+			break
+		}
+	}
+	if found { // append to most recent bucket in existing index
+		if err := b.Append(event); err != nil {
+			return nil, err
+		}
+	} else {
+		if b == nil {
+			b = &BucketInfo{
+				ConditionId: event.ConditionId.Id,
+			}
+		}
+		if err := b.Create(); err != nil {
+			return nil, err
+		}
+		if err := b.Append(event); err != nil {
+			return nil, err
+		}
+	}
+	return &empty.Empty{}, nil
+}
+
+func (p *Plugin) GetAlertLog(ctx context.Context, ref *corev1.Reference) (*corev1.AlertLog, error) {
+	return nil, shared.AlertingErrNotImplemented
+}
+
+// https://stackoverflow.com/questions/18879109/subset-check-with-slices-in-go
+func isSubset(first, second []string) bool {
+	set := make(map[string]int)
+	for _, value := range second {
+		set[value] += 1
+	}
+
+	for _, value := range first {
+		if count, found := set[value]; !found {
+			return false
+		} else if count < 1 {
+			return false
+		} else {
+			set[value] = count - 1
+		}
+	}
+	return true
+}
+
+func hasLabels(item *alertingv1alpha.InformativeAlertLog, labels []string) bool {
+	return len(labels) == 0 || isSubset(labels, item.Condition.Labels)
+}
+
+func hasRange(item *alertingv1alpha.InformativeAlertLog, s *timestamppb.Timestamp, e *timestamppb.Timestamp) bool {
+	if item.Log.Timestamp == nil { // shouldn't happen, but hey
+		return true
+	}
+	start := s.AsTime()
+	end := s.AsTime()
+	if s == nil && e == nil {
+		return true
+	} else if s == nil { // before end
+		t := item.Log.Timestamp.AsTime()
+		return t.Before(end)
+	} else if e == nil {
+		t := item.Log.Timestamp.AsTime()
+		return t.After(start)
+	} else {
+		t := item.Log.Timestamp.AsTime()
+		return t.After(start) && t.Before(end)
+	}
+}
+
+func (p *Plugin) ListAlertLogs(ctx context.Context, req *alertingv1alpha.ListAlertLogRequest) (*alertingv1alpha.InformativeAlertLogList, error) {
+	if p.inMemCache != nil {
```

Comment:
You should invert this if-statement to decrease indentation level

Reply:
Yep